### PR TITLE
Shopping Cart

### PR DESCRIPTION
**Pull request by Elrhino**

### CODE REVIEW

File: client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java [hunk 1]
```diff
@@ -44,18 +47,25 @@
...
+    @UiField
+    SpanElement numberOfItems;
+    @UiField
+    Element cartIcon;
 
     private final AppResources resources;
-
     private final TokenFormatter formatter;
     private final PlaceManager placeManager;
+    private Boolean shoppingBagOpen;
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java:72]: nl

File: client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBag.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.arcbees.beeshop.client.application;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBag.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java [hunk 3]
```diff
@@ -87,6 +101,21 @@ public void f(Element element) {
             }
         });
 
+        $(cartButton).click(new Function() {
+            @Override
+            public void f() {
+                if (shoppingBagOpen) {
+                    $(shoppingBagWidget).show();
+                    $(cartIcon).attr("class", "iconClose");
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java]: isn't the `iconXXX` style in a Resource file? If so you should reference that instead.

File: client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBagImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package com.arcbees.beeshop.client.application;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBagImpl.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java [hunk 5]
```diff
@@ -87,6 +101,21 @@ public void f(Element element) {
...
+                    shoppingBagOpen = false;
+                } else {
+                    $(shoppingBagWidget).hide();
+                    $(cartIcon).attr("class", "iconCart");
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/ApplicationView.java]: same here

File: client/src/main/java/com/arcbees/beeshop/client/application/ShoppingBagItem.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.arcbees.beeshop.client.application;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/ShoppingBagItem.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBagImpl.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.arcbees.beeshop.client.events.ShoppingBagChangedEvent;
+import com.google.gwt.event.shared.GwtEvent;
+import com.google.gwt.event.shared.HasHandlers;
+import com.google.inject.Inject;
+import com.google.web.bindery.event.shared.EventBus;
+
+public class CurrentShoppingBagImpl implements CurrentShoppingBag, HasHandlers {
+    private List<ShoppingBagItem> items = new ArrayList<>();
+    private final EventBus eventBus;
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/CurrentShoppingBagImpl.java:29]: nl

File: client/src/main/java/com/arcbees/beeshop/client/application/product/ProductPresenter.java [hunk 8]
```diff
@@ -51,11 +56,12 @@
             EventBus eventBus,
             MyView view,
             MyProxy proxy,
-            SharePanelPresenter sharePanel) {
+            SharePanelPresenter sharePanel,
+            CurrentShoppingBag currentShoppingBag) {
         super(eventBus, view, proxy, ApplicationPresenter.SLOT_MAIN);
+        this.currentShoppingBag = currentShoppingBag;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/product/ProductPresenter.java:65]: nl

Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/product/ProductPresenter.java:65]: nl

File: client/src/main/java/com/arcbees/beeshop/client/application/product/ProductPresenter.java [hunk 9]
```diff
@@ -41,9 +46,9 @@
     @NameToken(NameTokens.PRODUCTS)
     interface MyProxy extends ProxyPlace<ProductPresenter> {
     }
-
     static final PermanentSlot<SharePanelPresenter> SLOT_SHARE_PANEL = new PermanentSlot<>();
 
+    private final CurrentShoppingBag currentShoppingBag;
     private boolean isSharePanelShown;
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/product/ProductPresenter.java:54]: nl

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemFactory.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemFactory.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    private CurrentShoppingBag currentShoppingBag;
+
+    @Inject
+    ShoppingBagItemPresenter(
+            EventBus eventBus,
+            MyView view,
+            @Assisted ShoppingBagItem item,
+            CurrentShoppingBag currentShoppingBag) {
+        super(eventBus, view);
+        this.currentShoppingBag = currentShoppingBag;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java:45]: nl

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+        getView().setUiHandlers(this);
+        this.item = item;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java]: before with the other this?

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    @Inject
+    ShoppingBagItemPresenter(
+            EventBus eventBus,
+            MyView view,
+            @Assisted ShoppingBagItem item,
+            CurrentShoppingBag currentShoppingBag) {
+        super(eventBus, view);
+        this.currentShoppingBag = currentShoppingBag;
+
+        getView().setUiHandlers(this);
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemPresenter.java:48]: swap these 2 lines, and group the assignments

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemUiHandlers.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemUiHandlers.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemView.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagItemView.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagPresenter.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagPresenter.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagUiHandlers.java [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagUiHandlers.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEvent.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
+package com.arcbees.beeshop.client.events;
+
+import com.arcbees.beeshop.client.application.ShoppingBagItem;
+import com.google.gwt.event.shared.GwtEvent;
+import com.google.gwt.event.shared.HasHandlers;
+
+public class ShoppingBagChangedEvent extends GwtEvent<ShoppingBagChangedEventHandler> {
+    public static final Type<ShoppingBagChangedEventHandler> TYPE = new Type<>();
+    private final ShoppingBagItem item;
```
Comment (Pacane) [inline client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEvent.java:25]: format this

File: client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagView.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package com.arcbees.beeshop.client.application.widget;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/application/widget/ShoppingBagView.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEvent.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
+package com.arcbees.beeshop.client.events;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEvent.java:17]: copyright

File: client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEventHandler.java [hunk 22]
```diff
@@ -0,0 +1,7 @@
+package com.arcbees.beeshop.client.events;
```
Comment (olafleur) [inline client/src/main/java/com/arcbees/beeshop/client/events/ShoppingBagChangedEventHandler.java:17]: copyright

**Comment by Pacane** [comment]
LGTM after comment fixes + conflicts fixes

**Comment by olafleur** [comment]
LGTM

**Pull request merged**